fiddles with mech drills

Spookerton opened the pull request:
Refactors how mech drills interact with things.
Adds some material subtypes for drills and heads.
Examining a drill shows the state of the attached head.
Applies skills to changing heads on drills.

This hopefully fixes the infinite loops.

Spookerton commented:
appreciate a look from @CrimsonShrike

comma commented on code/modules/mechs/equipment/utility.dm:
```diff
@@ -296,122 +305,205 @@
+		owner.visible_message(SPAN_WARNING("[owner] revs the [drill_head] menancingly."))
+		playsound(src, 'sound/mecha/mechdrill.ogg', 50, 1)
+
+get_visible_durability
```
hurp

Spookerton replied:
haha, fug

CrimsonShrike commented on code/modules/mechs/equipment/utility.dm:
```diff
@@ -296,122 +305,203 @@
 		return
-	. = ..()
+	playsound(src, 'sound/weapons/circsawhit.ogg', 33, 1)
+	if (actor.unEquip(new_head, get_turf(src)))
```
Your actor can be the mech and not the user. I think here you want to unequip from user.

SierraKomodo replied:
See CrimsonShrike's comment regarding actor vs user

CrimsonShrike commented on code/modules/mechs/equipment/utility.dm:
```diff
@@ -296,122 +305,203 @@
+		if (5 to 10) delay = 2 SECONDS //plasteel, steel
+	owner.setClickCooldown(delay)
+
+	owner.visible_message(
```
Missing sound for drilling from what I can see, used to be playsound(src, 'sound/mecha/mechdrill.ogg', 50, 1)

SierraKomodo replied:
See CrimsonShrike's comment regarding the missing sound

SierraKomodo commented on code/modules/mechs/equipment/utility.dm:
```diff
@@ -263,23 +263,32 @@
-	to_chat(user, "It [descriptor].")
+	to_chat(user, "It [get_visible_durability()].")
+
+/obj/item/weapon/material/drill_head/steel/New(newloc)
```
Is there a particular reason to be using `New()` over `Initialize()` here?

Spookerton replied:
3am code probably.

Spookerton replied:
Oh, no. It's because that's the pattern for practically half the weapon/material subtypes in the codebase. I think I can use the other one easily enough here instead.

SierraKomodo commented on code/modules/mechs/equipment/utility.dm:
```diff
@@ -296,122 +305,203 @@
+	if (drill_head)
+		var/new_head_loc = new_head.loc
+		actor.visible_message(SPAN_ITALIC("\The [actor] begins to detach \the [drill_head] from \the [src]."), range = 5)
+		if (!do_after(actor, act_delay, drill_head, same_direction = user_pilot))
```
In this specific condition, there's no do_after progress bar, which made me think it wasn't working at first. Attaching the new head does show the progress bar.

SierraKomodo commented on code/modules/mechs/equipment/utility.dm:
```diff
@@ -296,122 +305,203 @@
+				continue
+			ore.Move(ore_box)
+
+/obj/item/mech_equipment/drill/attackby(obj/item/weapon/W, mob/user)
```
I'm not seeing any methods to remove an attached drill head that doesn't require using another drill head on it - Either via empty hand or with a tool. If you opt for a tool, I'd remove the ability to swap drill heads out by clicking on it with another head for consistency.

SierraKomodo commented on code/modules/mechs/equipment/utility.dm:
```diff
@@ -296,122 +305,203 @@
-		owner.visible_message("<span class='danger'>\The [owner] starts to drill \the [target]</span>", "<span class='warning'>You hear a large drill.</span>")
-		playsound(src, 'sound/mecha/mechdrill.ogg', 50, 1)
+	else if (istype(target, /turf/unsimulated))
+		to_chat(user, SPAN_WARNING("\The [target] can't be drilled away."))
```
This should have a `return`

SierraKomodo replied:
Still need a `return` here

SierraKomodo commented on code/modules/mechs/equipment/utility.dm:
```diff
@@ -296,122 +305,203 @@
+		SPAN_WARNING("\The [owner] starts to drill \the [target]."),
+		blind_message = SPAN_WARNING("You hear a large motor whirring.")
+	)
+	if (!do_after(owner, delay, target, same_direction = TRUE))
```
This call isn't showing a progress bar.

SierraKomodo reviewed (changes requested):
Ran through some local testing. See notes below and comments.

- It might be worth having the drill's name updated to include the drill head's material as a QOL update with this.

Spookerton committed:
mech drill tweaks & fix

generalizes drill head attaching into attach_head
adds premade material drill heads and drills for steel, titanium, plasteel, diamond & uses them
hopefully fixes a thing where sometimes drills would hard loop on the return of range(..) + corrects usage argument order

SierraKomodo reviewed (changes requested):
- I would like to be able to see what kind of drill head is installed when examining the drill - Either through the examine text and/or the drill's name itself. As it stands, it's not possible to tell what drill head is installed or even if one is installed outside of checking the integrity percentage in the mech HUD.

- I'd also like the durability to be visible on examining the drill itself - Currently this seems to only appear when examining a drill head.

Spookerton commented:
o ye

SierraKomodo commented:
There's still a couple of other items in the review comment that probably didn't show because GitHub does reviews weird:

- I would like to be able to see what kind of drill head is installed when examining the drill - Either through the examine text and/or the drill's name itself. As it stands, it's not possible to tell what drill head is installed or even if one is installed outside of checking the integrity percentage in the mech HUD.

- I'd also like the durability to be visible on examining the drill itself - Currently this seems to only appear when examining a drill head.

Spookerton commented:
@SierraKomodo Added as an addition to drill examine -- shows whether there is a drill head attached, and if less than 4 tiles away what material and condition the head is.
